draft-ietf-gnap-core-protocol.md

```diff
@@ -5365,3 +5365,693 @@ justification for the data type representing the same basic kind of thing it ext
 For example, an extension declaring an "array" representation for a field would need
 to explain how the array represents something akin to the non-array element that it
 is replacing. 
+
+# JSON Schema
+
+The following JSON Schemas are provided as a tool to help developers validate requests and
+responses within the GNAP protocol.
+
+Requesting access as in {{request}}:
+
+~~~
+{
+  "$schema": "https://json-schema.org/draft/2019-09/schema",
+  "name": "Request",
+  "definitions": {
+    "resource": {
+      "type": [
+        "string",
+        "object"
+      ],
+      "required": [
+        "type"
+      ],
+      "properties": {
+        "type": {
+          "type": "string"
+        },
+        "actions": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "locations": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "datatypes": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "identifier": {
+          "type": "string"
+        }
+      }
+    },
+    "key": {
+      "type": [
+        "object"
+      ],
+      "required": [
+        "proof"
+      ],
+      "properties": {
+        "jwk": {
+          "type": "object"
+        }/* this should have a JWK-specific subschema*/,
+        "cert": {
+          "type": "string"
+        },
+        "cert#S256": {
+          "type": "string"
+        },
+        "proof": {
+          "type": "string",
+          "enum": [
+            "jwsd",
+            "jws",
+            "mtls",
+            "dpop",
+            "pop",
+            "httpsig"
+          ]/* can be expanded by extensions*/
+        }
+      }
+    },
+    "subject_id_types": {
+      "type": "string",
+      "enum": [
+        "iss-sub",
+        "email",
+        "account",
+        "phone-number"
+      ]/* more added by extensions*/
+    },
+    "assertion_types": {
+      "type": "string",
+      "enum": [
+        "id_token",
+        "saml2"
+      ]/* more added by extensions*/
+    },
+    "subject_id_structures": {
+      "type": "object",
+      "required": [
+        "subject_type"
+      ],
+      "properties": {
+        "subject_type": {
+          "$ref": "#/definitions/subject_id_types"
+        }
+      },
+      "allOf": [
+        {
+          "if": {
```

Comment:
OMG. I think the more idiomatic way is to define an object for each option (you can use `const` for the `subject_type`), and then your resulting object "OneOf" them.

Reply:
There probably are better ways to set it up, and I'd be open to improvements. This is one of my arguments against mutual-exclusivity of object types like this within the protocol. This is the form of polymorphism that is painful to implement correctly. :)

Reply:
I disagree. JSON Schema has an clean idiomatic way to represent this stuff without all the "if/else", we just need to use it.